test/common/router/config_impl_test.cc

```diff
@@ -995,6 +995,99 @@ TEST(RouteMatcherTest, HeaderMatchedRouting) {
   }
 }
 
+TEST(RouteMatcherTest, QueryParamMatchedRouting) {
+  std::string json = R"EOF(
+{
+  "virtual_hosts": [
+    {
+      "name": "local_service",
+      "domains": ["*"],
+      "routes": [
+        {
+          "prefix": "/",
+          "cluster": "local_service_with_multiple_query_parameters",
+          "query_parameters": [
+            {"name": "id", "value": "\\d+[02468]", "regex": true},
+            {"name": "debug"}
+          ]
+        },
+        {
+          "prefix": "/",
+          "cluster": "local_service_with_query_parameter",
+          "query_parameters": [
+            {"name": "param", "value": "test"}
+          ]
+        },
+        {
+          "prefix": "/",
+          "cluster": "local_service_with_valueless_query_parameter",
+          "query_parameters": [
+            {"name": "debug"}
+          ]
+        },
+        {
+          "prefix": "/",
+          "cluster": "local_service_without_query_parameters"
+        }
+      ]
+    }
+  ]
+}
+  )EOF";
+
+  NiceMock<Runtime::MockLoader> runtime;
+  NiceMock<Upstream::MockClusterManager> cm;
+  ConfigImpl config(parseRouteConfigurationFromJson(json), runtime, cm, true);
+
+  {
+    Http::TestHeaderMapImpl headers = genHeaders("example.com", "/", "GET");
+    EXPECT_EQ("local_service_without_query_parameters",
+              config.route(headers, 0)->routeEntry()->clusterName());
+  }
+
+  {
+    Http::TestHeaderMapImpl headers = genHeaders("example.com", "/?", "GET");
+    EXPECT_EQ("local_service_without_query_parameters",
+              config.route(headers, 0)->routeEntry()->clusterName());
+  }
+
+  {
+    Http::TestHeaderMapImpl headers = genHeaders("example.com", "/?param=testing", "GET");
+    EXPECT_EQ("local_service_without_query_parameters",
+              config.route(headers, 0)->routeEntry()->clusterName());
+  }
+
+  {
+    Http::TestHeaderMapImpl headers = genHeaders("example.com", "/?param=test", "GET");
```

Comment:
Can you add a test for multiple query params?

Reply:
I put a couple of multi-param tests a bit later in config_impl_test.cc: [all parameters match](https://github.com/envoyproxy/envoy/pull/2258/commits/753d36950fa423561f4231a90a778c2aa236d49e#diff-fa2430fa87027867f55fc07dc4412eb3R1085) and [proper subset of parameters match](https://github.com/envoyproxy/envoy/pull/2258/commits/753d36950fa423561f4231a90a778c2aa236d49e#diff-fa2430fa87027867f55fc07dc4412eb3R1079).